### PR TITLE
Grace period option to health checks.

### PR DESCRIPTION
**Pull request by elifa**
**- What I did**
Added the option `--start-period` to `HEALTHCHECK` in order to allow containers with startup times to have a health check configured based on their behaviour once started.

The `--start-period` flag defines a period during which health check results are not counted towards the maximum number of retries configured by the `--retries` flag. However, if a health check succeeds during the grace period, any failures from there on will be counted towards the retries.

The default is to use no start period, meaning no change to current behaviour.

Additionally the `run` flag `--health-start-period` has been added to the CLI to override the value from the build.

The need for this has been discussed in #26498 and #26664 and this is my suggestion of how to solve the use cases discussed there.

**- How I did it**
Based on how long it has been since the container start `Container.State.StartedAt` and the given `--start-period`, it is determined whether the `Container.State.Health.FailingStreak` is incremented or not.

**- How to verify it**
A test has been added for verification of the new functionallity.

**- Description for the changelog**

Add `--start-period` flag to `HEALTHCHECK` and `run` override flag `--health-start-period` in order to enable health check for containers with an initial startup time.

Signed-off-by: Elias Faxö <elias.faxo@gmail.com>

### CODE REVIEW

**Comment by cpuguy83** [comment]
I don't really care for the name, but otherwise SGTM.
Maybe `start-period` to make it clear this is to give time during container startup, not once it's already started.

**Comment by elifa** [comment]
@cpuguy83 I agree, the name could be better. I'll change it to `start-period` instead and edit the PR.

File: docs/reference/builder.md
Comment (thaJeztah) [inline docs/reference/builder.md]: nit: typo s/retires/retries/

**Comment by thaJeztah** [comment]
I can see this being useful for services that have a long startup period, so sgtm. We can bike-shed over naming during review

**Comment by thaJeztah** [comment]
ping @aaronlehmann could you have a look if this would work for SwarmKit?

**Comment by aaronlehmann** [comment]
It would need to be plumbed through SwarmKit. This would involve a PR to SwarmKit to add the new flag to the protobuf definitions, and changes in Docker to add the flag to service create/update, and pass it through to the container in the executor.

**Comment by elifa** [comment]
@aaronlehmann Do you want me to add a PR to SwarmKit and add mapping for service to this PR? Was a bit uncertain if you wanted this to be merged before adding it to SwarmKit as it has a dependency to the types updated in this PR.

**Comment by aaronlehmann** [comment]
Not sure I'm the best one to answer that question. I think if maintainers are happy with the design of this PR, the next step would be to open a SwarmKit PR to add support there. But I don't want to get ahead of the design review.

**Comment by kakawait** [comment]
What do you think about a *readiness* healthcheck/probe in addition to `--health-start-period` in order to optimize `start-period` duration? 

If *readiness* healthcheck/probe returns `OK` before `--health-start-period` we can change state to `unhealthy` (then `healthy` when `healthcheck` will be triggered). However `--health-start-period` must be the max time, even if *readiness* healthcheck/probe always returns `NOT OK` (act as `timeout`).

File: daemon/health.go
Comment (dongluochen) [inline daemon/health.go]: State change should only happen when `FailingStreak` increases. It's better to move `if h.FailingStreak >= retries` block into `if shouldIncrementStreak`.

File: docs/reference/builder.md
Comment (dongluochen) [inline docs/reference/builder.md]: **start-period** provides initialization time for containers that need time to bootstrap. Probe failure during ...

**Comment by dongluochen** [comment]
Design sounds good to me.

**Comment by krasi-georgiev** [comment]
I think readiness test is more flexible

**Comment by aluzzardi** [comment]
My 2 cents: The title of the PR says *Grace period*. I kinda like that term better than *start period*, e.g. `--healthcheck-grace-period`.

@dongluochen ?

**Comment by dongluochen** [comment]
I don't have a preference between `grace-period` and `start-period`.

**Comment by thaJeztah** [comment]
We were discussing this in the maintainers meeting, and think this needs more discussion (healthcheck vs readiness check)

**Comment by dongluochen** [comment]
@thaJeztah Can you summarize the alternatives and how are we going to resolve?

**Comment by dnephin** [comment]
I believe it is well summarized in this comment: https://github.com/docker/docker/issues/26664#issuecomment-274451239

A fixed time interval is not a good solution for readiness checks, and there doesn't seem to be any other use cases for a grace period.

**Comment by thaJeztah** [comment]
We've been discussing this PR, and considering all options, this seems to be the best way forward. Although a "fixed time" can be hard to determine (e.g., how long does a database migration take?), setting a very long time (1 day) will work, because the "grace period" automatically "expires" once the service becomes healthy.

Having a separate "readiness" check doesn't fit in the SwarmKit design (@dongluochen will be better at explaining this).

Some things that need to be discussed is to create a list of service options that can be updated on a _running_ service, without causing a re-deploy of tasks (e.g., being able to _remove_ the grace period from the service definition)

I am moving this to code review but would like @dongluochen and @aluzzardi to check if this matches what we discussed

**Comment by dongluochen** [comment]
Today healthcheck is used in 2 different scenarios. The first scenario is `docker run`. Healthcheck result would move container state from `starting` to `healthy/unhealthy`. It doesn't take any action on the container. Healthcheck result/events are consumed by operators or external tools. 

The other scenario is in swarm mode where healthcheck serves 2 purposes. If a task is unhealthy, manager would shut it down and restart it (liveness check). The second purpose is to add the container to service records (load balancer and DNS) only when it becomes healthy (readiness check). 

There may be different requirements between liveness check and readiness check. A liveness check may be run once per minute to validate application running, while readiness check may be done every second to reduce request loss. There are cases you want different service endpoints. I have seen systems that separate these 2 checks or use a single check. In systems separating the 2 checks, a lot of applications would use the same endpoint for both checks, i.e., both checks visit the same service endpoint. Adding service endpoint to a container is relatively expensive. Liveness check needs a start grace period to allow application initialization. Users need to guess a upper limit of start time.

Since swarm mode uses healthcheck for both purposes now, we need a start grace period. In the future if we decide to separate them, we would add readiness check for service records to provide flexibility in load balancing. I don't see adding start grace period has conflict of future development.

**Comment by cpuguy83** [comment]
It why add it if the purpose is to serve a case that would be much better served by something else?

**Comment by dongluochen** [comment]
> why add it if the purpose is to serve a case that would be much better served by something else?

If we are moving to the liveness/readiness model, start grace period for liveness check is still required. I don't see conflicts. Do I miss something?

**Comment by elifa** [comment]
@dongluochen made some updates according to your comments.

When it comes to liveness/readiness, I cannot see that this PR would be in conflict with a vision to enable separation of liveness and readiness verification endpoints other than that this feature would only live on in the declaration of the liveness check.

**Comment by cpuguy83** [comment]
@dongluochen I don't see why it would be required if it's split? A "grace period" would be simply timeout period * number of retries.
The reason to add grace period here is to account for startup before flagging which would not be needed if we have a check specifically for that?

**Comment by krasi-georgiev** [comment]
I think the same it should be one or the other

**Comment by dongluochen** [comment]
> I don't see why it would be required if it's split? A "grace period" would be simply timeout period * number of retries.

@cpuguy83 I think "start grace period" is still required for liveness check even it's split from readiness check. Let's assume a container takes ~1 hour to start because it has to validate data integrity before responding. You want liveness check to be meaningful, e.g., do a query/response validation. Here liveness check wouldn't be successful in the first hour at start. You don't want `timeout * number of retries` to stretch to 1 hour. `timeout * number of retries` is the normal failure detection time. It's directly related to Mean Time To Recovery (MTTR) because it triggers task repair (restart). This should be fast, in the range of several seconds or 1 minute. 

Theoretically you can inject a liveness check into your image so it starts responding right after start, which can confirm the task is progressing properly. In practice it requires a lot of effort and hard to get right.

**Comment by cpuguy83** [comment]
@dongluochen Then I'm not sure what the purpose of the split would be.

**Comment by dongluochen** [comment]
The split gives flexibility in controlling task restart and request routing. Liveness check is wired into task control and readiness check decides if it should take requests. Let's say a container is overloaded and starts failing both requests. The readiness check can be `(1 second timeout + 1 second interval) * 3 retries` so the load balancer removes the container from service rotation after 6 seconds. The liveness check could be `(1 second timeout + 10 second interval) * 12 retries = 132 seconds`. If the container recovers within 2 minutes, the container would not be killed. And it's not serving requests before recovery. 

I think the current healthcheck doing both checks is acceptable and may fit most use cases. If users think they should be separated, a readiness check may be added. There is a cost associated with each check.

**Comment by krasi-georgiev** [comment]
@dongluochen - a bit confusing with the terminology. could you maybe give some examples to simplify your idea and use cases?

**Comment by dongluochen** [comment]
@krasi-georgiev The initial plan for healthcheck https://github.com/docker/docker/issues/21142 includes "grace period". You should find most terminology there.

**Comment by krasi-georgiev** [comment]
thanks, I understand the idea for grace period and this will solve some problems in the images I maintain, but don't get what do you mean by  liveness, readiness checks in addition to grace period ?

**Comment by dongluochen** [comment]
Liveness check in this context is how the system decides if a container is running as expected, e.g., not in deadlock/livelock. The system restarts a container if liveness check fails. 

Readiness check decides if a task (container) is ready to accept requests. The container is added to load balancer/DNS record when it passes readiness check, and removed when it fails. 

Current healthcheck is used as liveness check and readiness check.

**Comment by dongluochen** [comment]
@cpuguy83 @justincormack @aluzzardi Any other ideas of this design? If not, we would move forward to code review.

**Comment by cpuguy83** [comment]
@dongluochen This is not how I see it.

A readiness check is the check that is run while the container is in it's startup period.
It gets run until either retries*timeout is reached or returns success (that it was started and is ready).
From there liveness kicks in and should pretty much behave as it does today... only difference being we know the process should be fully ready to take requests.

However I suppose we can add this for 1 use-case and potentially add the readiness check later for other use-cases.

**Comment by krasi-georgiev** [comment]
@cpuguy83 very well summarised!

on a side note using readiness/liveness  for the names is a bit confusing and I am sure it would cause a lot of questions. I think we should keep healthcheck and choose another name for the readiness

**here are the suggestions so far.** 
bootstrap-period +1
grace-period
start-period

**Comment by dongluochen** [comment]
@cpuguy83 I think we are trying to address same problem with different approaches. The `readiness` in your context looks to solve the "start grace period" problem. 

Orchestration system needs periodic checks to decide if a task is alive or not throughout task life cycle. Load balancer relies on periodic checks to decide if a container can serve traffic or not. In many use cases load balancer is external and I think we need to support them, maybe with events. Using the same check for orchestrator and load balancer as of today is simple and fits most use cases. I think adding start grace period like this PR would be able to solve the long startup problem, without making extra changes.

**Comment by dongluochen** [comment]
@elifa Can you update code to fix janky. One error I see is:

```
01:40:45 These files are not properly gofmt'd:
01:40:45  - daemon/health.go
01:40:45 
01:40:45 Please reformat the above files using "gofmt -s -w" and commit the result.
```

**Comment by elifa** [comment]
@dongluochen I have fixed the formatting error now.

**Review by dongluochen — CHANGES_REQUESTED** [review]
You also need to update `cli/command/service/opts.go` and add flag in service create/update.

**Comment by elifa** [comment]
@dongluochen I have updated the opts and run-documentation.

I havn't added mapping for service since it would require a PR to SwarmKit to map the `start-period` field through there, and that would require these changes to be merged first as SwarmKit depends on the types here. Or would you like me to add the field there anyway?

**Comment by thaJeztah** [comment]
> and that would require these changes to be merged first as SwarmKit depends on the types here. Or would you like me to add the field there anyway?

You can make the changes to the vendored SwarmKit files here first (as a separate commit). The `vendor` CI check will fail, but we can ignore that during review. Doing so allows us to review all changes here, and if we're ok with the changes, you can open a pull request against the SwarmKit repository, and "bump" the vendored files here after that's merged.

**Comment by dongluochen** [comment]
@elifa I think we agree on your current approach. Can you add corresponding PR to Swarmkit? Then we can vendor it in to complete this PR.

**Comment by elifa** [comment]
Sorry about the delay @dongluochen. The mapping for service is added now and a PR is created to SwarmKit https://github.com/docker/swarmkit/pull/2044

Both PRs have a commit manually adding the updated types to the vendor directory.

**Comment by thaJeztah** [comment]
Looks like CI is failing;

```
19:41:53 Building: bundles/17.05.0-dev/binary-client/docker-17.05.0-dev
19:42:13 # github.com/docker/docker/cli/command/container
19:42:13 cli/command/container/opts.go:488: too many arguments to return
19:42:14 Build step 'Execute shell' marked build as failure
```

Also note that there's a separate vendor PR https://github.com/docker/docker/issues/32010, which may bring in the required changes

**Comment by elifa** [comment]
@thaJeztah Thanks, I saw that other PR right after i pushed. I'll just wait for that one to be merged then.

**Comment by thaJeztah** [comment]
Looks like there's a couple of PRs blocking each other (see the discussion on the other PR), so we need some strategy haha

File: cli/command/container/opts.go
Comment (allencloud) [inline cli/command/container/opts.go]: since docker daemon would expose this in the api, thus I think we need to update this parameter in the daemon side as well, WDYT? @elifa

Comment (elifa) [inline cli/command/container/opts.go]: You're absolutely right. I'll add the same validation on the daemon side.

**Comment by thaJeztah** [comment]
@elifa https://github.com/docker/docker/pull/32010 was merged, so you should be able to rebase to get the SwarmKit version bump

File: api/types/container/config.go
Comment (teohhanhui) [inline api/types/container/config.go]: retires -> retries

Comment (elifa) [inline api/types/container/config.go]: Thanks, I corrected the misspelling :)

**Comment by elifa** [comment]
@thaJeztah Rebased and ready :)

File: cli/command/container/opts.go
Comment (thaJeztah) [inline cli/command/container/opts.go]: Can you add an annotation here so that it's hidden when talking to an older daemon (API version for [Docker 17.05 will be 1.29](https://github.com/docker/docker/blob/817106aaa1e8575d86e0b95a4219f6a7434572e3/api/common.go#L23-L24))?

```go
flags.SetAnnotation("health-start-period", "version", []string{"1.29"})
```

File: cli/command/service/opts.go
Comment (thaJeztah) [inline cli/command/service/opts.go]: This version now should be `1.29`

File: cli/command/service/opts.go
Comment (thaJeztah) [inline cli/command/service/opts.go]: @dongluochen should this be s/container/service/ ?

File: api/types/container/config.go
Comment (thaJeztah) [inline api/types/container/config.go]: Can you update the swagger.yml to include the new field? https://github.com/docker/docker/blob/9f9cc221bdb0a919d6ed28bdfaffa9ae2fe3a5fd/api/swagger.yaml#L485-L508

Also can you add a line to the API version-history; https://github.com/docker/docker/blob/master/docs/api/version-history.md#v129-api-changes

**Review by thaJeztah — CHANGES_REQUESTED** [review]
Thanks @elifa, I quickly went through the changes and saw some minor nits, but I think we're almost there!

File: cli/command/container/opts.go
Comment (elifa) [inline cli/command/container/opts.go]: The annotation is added.

File: cli/command/service/opts.go
Comment (elifa) [inline cli/command/service/opts.go]: Version is changed to `1.29` now.

File: api/types/container/config.go
Comment (elifa) [inline api/types/container/config.go]: The field has been added and the line in the version-history as well.

**Review by thaJeztah — APPROVED** [review]
LGTM, thanks!

**Comment by thaJeztah** [comment]
ping @dongluochen PTAL

**Comment by thaJeztah** [comment]
also /cc @albers @sdurrheimer :smile:

File: api/swagger.yaml
Comment (dongluochen) [inline api/swagger.yaml]: > It should be 0 or not less than 1000000000(1s).

Any specific reason for this constraint? I think `StartPeriod` can be any non-negative value.

File: cli/command/service/opts.go
Comment (dongluochen) [inline cli/command/service/opts.go]: It should be container.

File: api/swagger.yaml
Comment (elifa) [inline api/swagger.yaml]: I'll change it to any non-negative.

File: daemon/container.go
Comment (dongluochen) [inline daemon/container.go]: > if config.Healthcheck.StartPeriod != 0 && config.Healthcheck.StartPeriod < 0 {

if config.Healthcheck.StartPeriod < 0 {

Comment (elifa) [inline daemon/container.go]: Let's pretend that never happened .. ;)

**Review by dongluochen — APPROVED** [review]
LGTM

**Comment by thaJeztah** [comment]
Gonna take @dongluochen's LGTM into account here. All green 👍

**Comment by thaJeztah** [comment]
Thanks so much @elifa !

**Comment by elifa** [comment]
Great! Thanks for all your help!
I have created a PR towards swarmkit https://github.com/docker/swarmkit/pull/2103 finalizing the feature.

File: docs/api/version-history.md
Comment (ijc25) [inline docs/api/version-history.md]: "unealthy" is a typo I think.

Comment (albers) [inline docs/api/version-history.md]: @ijc25 created #32523 for that

**Comment by pascalandy** [comment]
Hello gents,
	I'm so glad to see appearing. Since the healthcheck, service take 30s to 180s to deploy.
	
	I'm not sure to understand the difference between the two flags.
	I don't want any healthcheck for the first 90 seconds. I guess I only need to use:
	
	`--health-start-period "90s"`
	
	I'm not sure why `--start-period "90s"` exist. Can you enlighten me?
	
	Here is my core setup to start nginx:
	
	
	```
docker service create \
	--name "$CTN_nginx_app" \
	--hostname "$CTN_nginx_app" \
	--network "$NTW_FRONT" \
	--replicas "1" \
	--reserve-memory "12M" \
	--limit-memory "20M" \
	--constraint node.labels.apps_accepted=="yes" \
	--mount	type="bind",src="$WWW_SRC_NGINX",target="$WWW_DST_NGINX" \
	--restart-condition "any" \
	--restart-max-attempts "55" \
	--update-delay "5s" \
	--update-parallelism "1" \
	--start-period "90s" \
	--health-start-period "90s" \
nginx:alpine
	```
Cheers!
Pascal

**Comment by thaJeztah** [comment]
@pascalandy They're the same; the `--start-period` is not a command-line option, but an option to the `HEALTHCHECK` Dockerfile instruction to define the start-period as part of the image, whereas `--health-start-period` is a command-line option to set/override that period at runtime.

This option is only used if the image / container you're running uses a health check (your example uses the `nginx:alpine` image, which does not have a health check defined). When using the `--health-start-period`, it works roughly like this;

Say, the service is created with

```bash
$ docker service create --name=health \
  --health-cmd='exit 1' \
  --health-interval=10s \
  --health-timeout=3s \
  --health-retries=3 \
  --health-start-period=60s \
  nginx:alpine
```

1. The container (task) is created and started
2. Every `10 seconds` the command `foobar` is executed to check if the container is healthy
3. If the health-cmd, `exit 1`, fails (hint: that's _always_ :smile:), but the container has not been running for more than `60 seconds` (the start period), don't count the failure, and let the container run as usual
4. After `60 seconds`, start tracking failures. If `exit 1` fails 3 times in a row, mark the container/task as "unhealthy", stop it, and start a new task to replace it.

If you run the above example, you can see follow what's happening. During the first 60 seconds, you can inspect the container, and see that the health check is failing, and failures are logged;


```bash
$ docker inspect --format '{{json .State.Health }}' be45d670f023 | jq .
{
  "Status": "starting",
  "FailingStreak": 0,
  "Log": [
    {
      "Start": "2017-05-16T09:53:21.308489492Z",
      "End": "2017-05-16T09:53:21.360249491Z",
      "ExitCode": 1,
      "Output": ""
    }
  ]
}

$ docker inspect --format '{{json .State.Health }}' be45d670f023 | jq .
{
  "Status": "starting",
  "FailingStreak": 0,
  "Log": [
    {
      "Start": "2017-05-16T09:53:21.308489492Z",
      "End": "2017-05-16T09:53:21.360249491Z",
      "ExitCode": 1,
      "Output": ""
    },
    {
      "Start": "2017-05-16T09:53:31.361873326Z",
      "End": "2017-05-16T09:53:31.415394581Z",
      "ExitCode": 1,
      "Output": ""
    }
  ]
}
```

However, _even though it fails 3 times or more in a row_, the `FailingStreak` remains `0`, and `Status` remains `"starting"` (because we're still in the "start period");

```bash
$ docker inspect --format '{{json .State.Health }}' be45d670f023 | jq .
{
  "Status": "starting",
  "FailingStreak": 0,
  "Log": [
    {
      "Start": "2017-05-16T09:53:21.308489492Z",
      "End": "2017-05-16T09:53:21.360249491Z",
      "ExitCode": 1,
      "Output": ""
    },
    {
      "Start": "2017-05-16T09:53:31.361873326Z",
      "End": "2017-05-16T09:53:31.415394581Z",
      "ExitCode": 1,
      "Output": ""
    },
    {
      "Start": "2017-05-16T09:53:41.415861379Z",
      "End": "2017-05-16T09:53:41.452461489Z",
      "ExitCode": 1,
      "Output": ""
    },
    {
      "Start": "2017-05-16T09:53:51.453103536Z",
      "End": "2017-05-16T09:53:51.490980125Z",
      "ExitCode": 1,
      "Output": ""
    },
    {
      "Start": "2017-05-16T09:54:01.492147629Z",
      "End": "2017-05-16T09:54:01.533664526Z",
      "ExitCode": 1,
      "Output": ""
    }
  ]
}
```

Once the container is running for the `60 seconds`, docker starts to track failures (`FailingStreak` is incremented with each consecutive failure);

```bash
$ docker inspect --format '{{json .State.Health }}' be45d670f023 | jq .
{
  "Status": "starting",
  "FailingStreak": 1,
  "Log": [
    {
      "Start": "2017-05-16T09:53:31.361873326Z",
      "End": "2017-05-16T09:53:31.415394581Z",
...
```

And when it reaches `--health-retries` (3), the container/task is marked as unhealthy;

```bash
$ docker inspect --format '{{json .State.Health }}' be45d670f023 | jq .
{
  "Status": "unhealthy",
  "FailingStreak": 3,
  "Log": [
    {
      "Start": "2017-05-16T09:53:51.453103536Z",
      "End": "2017-05-16T09:53:51.490980125Z",
...
```

At that point, Swarm takes control; stops the container/task and starts a new one to replace it;

```bash
$ docker service ps health
ID                  NAME                IMAGE               NODE                DESIRED STATE       CURRENT STATE                 ERROR               PORTS
uqaf5bbiunnf        health.1            nginx:alpine        91d5d251ecc3        Running             Starting 20 seconds ago
be45d670f023         \_ health.1        nginx:alpine        91d5d251ecc3        Shutdown            Complete 25 seconds ago
```

**Comment by thaJeztah** [comment]
To add to the above; the "health start period" (or "grace period", which is also used as a term), allows you to monitor a service's health (by inspecting the `.State.Health` of the container), during the startup period. This can be helpful for services that need to perform certain tasks the first time they are started (think of a database-migration), but you still want to keep track if the migration is running, and if the health-checks are running (you can log messages as part of the health check, which show up in `.State.Health` in the container inspect).

Swarm mode takes the health-state into account when routing network traffic to the task. The startup period can take _less_ than the specified amount (e.g. the database migration took less time than expected, whoop!), at which point the container/task will start to receive network requests.


To illustrate the above; a simple example; the health-check below simulates a long-running startup. The first 40 seconds (4 healthcheck intervals), the healthcheck returns "unhealthy". Because of the `health-start-period` the container is not terminated, but the health-check is still performed every 10 seconds;

```bash
$ docker service create --name=health \
  --health-cmd='if [ ! -f "/count" ] ; then ctr=0; else ctr=`cat /count`; fi; ctr=`expr ${ctr} + 1`; echo "${ctr}" > /count; if [ "$ctr" -gt 4 ] ; then exit 0; else exit 1; fi' \
  --health-interval=10s \
  --health-timeout=3s \
  --health-retries=3 \
  --health-start-period=60s \
  -p8080:80 \
  nginx:alpine
```

As long as the container is not "healthy", no traffic is routed to the task;

```bash
$ curl localhost:8080
curl: (7) Failed to connect to localhost port 8080: Connection refused
```

After 40 seconds, the container becomes healthy, and Swarm starts to route traffic to it;

```bash
$ curl localhost:8080

<!DOCTYPE html>
<html>
<head>
<title>Welcome to nginx!</title>
<style>
    body {
```

**Comment by pascalandy** [comment]
This is crystal clear now @thaJeztah. Thank you so much for this deep explanation :)

**Comment by thaJeztah** [comment]
@pascalandy you're welcome! I took a bit of time to write it down, because I noticed that documentation around this was largely missing, so thought it would help as a starting point for that (I opened an issue in the documentation repository: https://github.com/docker/docker.github.io/issues/3282).

To come back to your initial comment;

> my services take 30s to 180s to deploy

Be aware that the _deploy_ time is separate from the "start period"; the deploy time may include _pulling_ the image before the task/container is started. This time is _not_ part of the start period (which starts once the task/container is actually started).

**Comment by pascalandy** [comment]
Perfectly aware but the pulling is already done.

> the deploy time may include pulling the image before the task/container is started

**Comment by vide** [comment]
@thaJeztah is this already integrated with Compose v3.X? Or should I open a specific issue to have it implemented?

**Comment by thaJeztah** [comment]
@vide a quick glance at the docker-compose 3.3 schema tells me it's not implemented yet; https://github.com/docker/cli/blob/master/cli/compose/schema/data/config_schema_v3.3.json#L310-L326

Can you open an issue in the https://github.com/docker/cli/issues issue tracker?

**Comment by omerh** [comment]
I am on a fresh install of 17-06 stable
```
Server:
 Version:      17.06.0-ce
 API version:  1.30 (minimum version 1.12)
 Go version:   go1.8.3
 Git commit:   02c1d87
 Built:        Fri Jun 23 21:19:04 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

I dont have the option in CLI run command:
```
docker service update --health-
--health-cmd       --health-interval  --health-retries   --health-timeout
```
missing cli switch is --health-start-period in stable 17.16 ??

**Comment by xdmitry** [comment]
@thaJeztah is it possible to change non-swarm docker instances to behave similarly to swarm service?

i.e. in your example you can't access port 8080 before service becomes healthy

but if you do 
<pre>
docker run -d --name=health \
  --health-cmd='if [ ! -f "/count" ] ; then ctr=0; else ctr=`cat /count`; fi; ctr=`expr ${ctr} + 1`; echo "${ctr}" > /count; if [ "$ctr" -gt 4 ] ; then exit 0; else exit 1; fi' \
  --health-interval=10s \
  --health-timeout=3s \
  --health-retries=3 \
   --health-start-period=60s \
  -p 8080:80   nginx:alpine \
&& sleep 1 && curl -s localhost:8080 | head -2 \
&& docker inspect health --format '{{.State.Health.Status}}'
</pre>

you'll get something like
<pre>

818af365e18fb77448ec99babab260289452e030ac9f2f974867ac93ff81cf31
&lt;!DOCTYPE html&gt;
&lt;html&gt;
starting
</pre>

instead of 

<pre>
curl: (7) Failed to connect to localhost port 8080: Connection refused
<pre>

**Comment by thaJeztah** [comment]
@xdmitry not sure that would be possible easily 🤔 there's no reconciliation loop or monitor process on non-swarm containers. Is there a reason for not using a swarm service for your use-case?

**Comment by felipecrs** [comment]
The documentation does not make this clear. Is `--start-period` only valid for the first initilization of a container, or is it valid even when the container is restarted?

**Comment by felipecrs** [comment]
I just tested here. Docker takes into account `--start-period` every time the container starts, not only during the first initialization. That's a little odd, because, for example, the Oracle DB container takes 10 minutes to start for the first time (because it needs to create all the DB files), while subsequent starts takes less than 1 minute.

Any suggestion?

- Refs https://github.com/oracle/docker-images/issues/2462

**Pull request merged**
